src/core_modules/capture-core/metaDataStoreLoaders/programs/loadPrograms.js

```diff
@@ -1,110 +1,171 @@
-            if (optionSet) {
-                accProgramStageOptionSets.push(optionSet);
-            }
-            return accProgramStageOptionSets;
-        }, []);
-    };
-
-    const getProgramOptionSets = (program) => {
-        const programStages = program.programStages || [];
-        return programStages.reduce((accProgramStagesOptionSets, programStage) => [
-            ...accProgramStagesOptionSets, ...getProgramStageOptionSets(programStage),
-        ], []);
-    };
-
-    return programsOutline.reduce((accOptionSets, program) => [
-        ...accOptionSets, ...getProgramOptionSets(program),
-    ], []);
-}
-
-function getTrackedEntityAttributeIds(missingPrograms) {
-    return missingPrograms
-        ? missingPrograms.reduce((accAttributeIds, program) => {
-            if (program.programTrackedEntityAttributes) {
-                const attributeIds =
-                    program.programTrackedEntityAttributes
+    return storageController
+        .getAll(storeNames.PROGRAMS, {
+            project: program => ({
+                id: program.id,
+                version: program.version,
+            }),
+        });
+};
+
+/**
+ * Remove programs from the cache that wasn't retrieved from the api
+ * @param {Object[]} apiPrograms: programs from the api
+ * @param {Object[]} cachePrograms: programs from the cache
+ */
+const removeUnavailablePrograms = async (apiPrograms, cachePrograms) => {
+    const apiProgramsObject = apiPrograms
+        .reduce((acc, apiProgram) => {
+            acc[apiProgram.id] = apiProgram;
+            return acc;
+        }, {});
+
+    const unavailableProgramIds = cachePrograms
+        .filter(cacheProgram => !apiProgramsObject[cacheProgram.id])
+        .map(unavailableProgram => unavailableProgram.id);
+
+    if (unavailableProgramIds.length > 0) {
+        const { storageController, storeNames } = getContext();
+        await storageController.remove(storeNames.PROGRAMS, unavailableProgramIds);
+    }
+};
+
+/**
+ * Retrieve the program ids for the programs we would like to update in the cache
+ * @param {Object[]} apiPrograms: programs from the api
+ * @param {Object[]} cachePrograms: programs from the cache
+ */
+const getStaleProgramIds = (apiPrograms, cachePrograms) => {
+    const cacheProgramsAsObject = cachePrograms
+        .reduce((acc, cacheProgram) => {
+            acc[cacheProgram.id] = cacheProgram;
+            return acc;
+        }, {});
+
+    return apiPrograms
+        .filter((program) => {
+            const cachedProgram = cacheProgramsAsObject[program.id];
+            return !cachedProgram || cachedProgram.version !== program.version;
+        })
+        .map(program => program.id);
+};
+
+/**
+ * Update the cache for the program ids passed in.
+ * @param {string[]} programIds: the program ids with stale data in the cache
+ * @returns programs data that is needed for the side effects.
+ */
+const loadProgramBatch = async (programIds) => {
```

Comment:
Can you please explain to me what makes us not choosing to have a self invoking function here? I am asking so that I can better understand the paradigms we use and when :)